drivers/at24cxxx/at24cxxx.c

```diff
@@ -0,0 +1,366 @@
+/**
+ * @ingroup     drivers_at24cxxx
+ * @{
+ *
+ * @file
+ * @brief       Device driver implementation for the AT24C128 / AT24C256 EEPROM units.
+ *
+ * @}
+ */
+
+#include <stdio.h>
+#include <errno.h>
+#include <string.h>
+
+#define ENABLE_DEBUG  (0)
+#include "debug.h"
+#include "assert.h"
+#include "xtimer.h"
+
+#include "include/at24cxxx_params.h"
+#include "include/at24cxxx_defines.h"
+#include "at24cxxx.h"
+
+/*
+    page 0: 0-63
+    page 1: 64-127
+    page 2: 128-191
+    page 3: 192-255
+    page 4: 256-319
+    page 5: 320-383
+    page 6: 384-447
+    page 7: 448-511
+    page 8: 512-575
+    ...
+ */
+static inline void _position_to_address_words(uint8_t addr[AT24CXXX_ADDR_LEN],
+                                              uint32_t pos)
+{
+    uint16_t a = pos & 0xFFFF;
+
+    addr[0] = (a & 0xFF00) >> 8;
+    addr[1] = a & 0x00FF;
+}
+
+static int _read_byte(const at24cxxx_t *dev, uint32_t pos)
+{
+    uint8_t buffer[AT24CXXX_ADDR_LEN];
+
+    _position_to_address_words(buffer, pos);
+
+    int check;
+
+    //dummy write
+    check = i2c_write_bytes(dev->params.i2c, dev->params.dev_addr, buffer,
+                            sizeof(buffer), I2C_NOSTOP);
+    DEBUG("write_byte check: %d\n", check);
+    if (check < 0) {
+        //check is a negative errno value
+        return check;
+    }
+
+    //current address read
+    check = i2c_read_byte(dev->params.i2c, dev->params.dev_addr, buffer, 0);
+    DEBUG("read_byte check: %d\n", check);
+    if (check < 0) {
+        //check is a negative errno value
+        return check;
+    }
+
+    return *buffer;
+}
+
+static int _read(const at24cxxx_t *dev, uint32_t pos, uint8_t *data, size_t len)
+{
+    uint8_t buffer[AT24CXXX_ADDR_LEN];
+
+    _position_to_address_words(buffer, pos);
+
+    int check;
+
+    //dummy write
+    check = i2c_write_bytes(dev->params.i2c, dev->params.dev_addr, buffer,
+                            sizeof(buffer), I2C_NOSTOP);
+    DEBUG("write check: %d\n", check);
+    if (check < 0) {
+        //check is a negative errno value
+        return check;
+    }
+
+    //sequential read
+    check = i2c_read_bytes(dev->params.i2c, dev->params.dev_addr, data, len, 0);
+    DEBUG("read check: %d\n", check);
+    if (check < 0) {
+        //check is a negative errno value
+        return check;
+    }
+    return len;
+}
+
+static int _write_byte(const at24cxxx_t *dev, uint32_t pos,  uint8_t data)
+{
+    uint8_t buffer[AT24CXXX_ADDR_LEN + 1];
+
+    _position_to_address_words(buffer, pos);
+
+    buffer[sizeof(buffer) - 1] = data;
+
+    int check;
+
+    check = i2c_write_bytes(dev->params.i2c, dev->params.dev_addr, buffer,
+                            sizeof(buffer), 0);
+    DEBUG("write_byte check: %d\n", check);
+    if (check < 0) {
+        //check is a negative errno value
+        return check;
+    }
+
+    xtimer_usleep(AT24CXXX_MAX_WRITE_CYCLE * US_PER_MS);
```

Comment:
Does it really always take 5ms to complete a write?
You can also use polling to determine if a write is finished:

>  Once the internally-timed write cycle has started and the
EEPROM inputs are disabled, acknowledge polling can be initiated. This involves sending a
start condition followed by the device address word. The read/write bit is representative of the
operation desired. Only if the internal write cycle has completed will the EEPROM respond
with a zero, allowing the read or write sequence to continue.

If I understand it correctly, polling until `i2c_read_byte()` returns 0 (no error) again should work.

Reply:
Some delay after a write operation was necessary.
A better approach would be to attempt to perform the requested operation for at most 5ms, right? But I think I would poll for the return value not being -ETIMEDOUT, instead of being 0.

Reply:
This seems to be a power consumption vs access time tradeoff to me. Continuously polling the I2C bus for up to 5 ms will keep the bus and the MCU busy, but could potentially lead to faster access times (when the write cycle is completed faster than the 5ms).

Also, I remeber from writing the `i2c_scan` shell command that "pinging" the device does not work reliable on all I2C hardware implementations. (If I remeber correctly, some MCUs required to actually read or write one byte of data, as the subsequent transfer always failed when only the address but now data was transferred.) So just waiting might also be more reliable? @MrKevinWeiss has superior insight into all the gory details due to his testing effort (afaik), maybe he can confirm or dismiss my doubt?

If waiting without polling will "win", I'd suggest to release the bus before the wait and acquire it again afterwards. That should allow the MCU to power down the bus or let other drivers access the potentially shared bus. (E.g. in I2C standard speed (100kbps) about 60 bytes of data could be transferred within the 5ms, so this can be considered as rather long.)

I'd personally prefer the worst case wait (with bus powered down / available for other tasks) over faster access time.

Reply:
I'd say a 1ms delay between each polling wold be a possible compromise. What I like about it is that we could do away with a device specific constant and have it 'just work' with any EEPROM than expects a different delay, be it longer or shorter.

Reply:
Quite a few devices will read one byte automatically if the address is ACKed, this means you cannot reliably only poll the address.  I believe this is only for the read as I don't remember having these problems in the write address.  I also _think_ you should not be allowed to write or read regs with length 0 though.

If you wanted to only exit after everything is free (so blocking) then I think you can just  try to read one byte `i2c_read_byte()` and discard it (or even compare to make sure the bytes are written). I think it should actually give you either a address nack or a data nack...  Here is the section.

> ACKNOWLEDGE POLLING: Once the internally-timed write cycle has started and the
EEPROM inputs are disabled, acknowledge polling can be initiated. This involves sending a
start condition followed by the device address word. The read/write bit is representative of the
operation desired. Only if the internal write cycle has completed will the EEPROM respond
with a zero, allowing the read or write sequence to continue.

The other option is, in the eeprom_write_byte/s have your retry if NACKed for 5ms.  If you do this you may also need to retry the read_bytes as I think it should also NACK a read command if still writing. This would mean you only poll if it is waiting and the poll wouldn't be too power hungry anyways since it would be a max of 5 polls.

The i2c error code should not be `-ETIMEDOUT` though.  It should be a NACK.

Reply:
When I try to write without a delay, the return type of e.g. `i2c_write_regs()` is `-ENXIO`, which indicates that

>  no devices responds on the address sent on the bus

Is this a NACK on the wire?

Reply:
Yup that is an address nack.  Good it is as expected.